src/PhpParser/ClassLikeAstResolver.php

```diff
@@ -27,51 +27,51 @@ final class ClassLikeAstResolver
     public function __construct(
         private readonly SmartPhpParser $smartPhpParser,
         private readonly BetterNodeFinder $betterNodeFinder,
+        private readonly NodeNameResolver $nodeNameResolver,
     ) {
     }
 
     public function resolveClassFromClassReflection(
-        ClassReflection $classReflection,
-        string $desiredClassName
+        ClassReflection $classReflection
     ): Trait_ | Class_ | Interface_ | Enum_ | null {
         if ($classReflection->isBuiltin()) {
             return null;
         }
 
-        if (isset($this->classLikesByName[$classReflection->getName()])) {
-            return $this->classLikesByName[$classReflection->getName()];
+        $className = $classReflection->getName();
+        if (isset($this->classLikesByName[$className])) {
+            return $this->classLikesByName[$className];
         }
 
         $fileName = $classReflection->getFileName();
 
         // probably internal class
         if ($fileName === null) {
             // avoid parsing falsy-file again
-            $this->classLikesByName[$classReflection->getName()] = null;
+            $this->classLikesByName[$className] = null;
             return null;
         }
 
         $stmts = $this->smartPhpParser->parseFile($fileName);
         if ($stmts === []) {
             // avoid parsing falsy-file again
-            $this->classLikesByName[$classReflection->getName()] = null;
+            $this->classLikesByName[$className] = null;
             return null;
         }
 
         /** @var array<Class_|Trait_|Interface_|Enum_> $classLikes */
         $classLikes = $this->betterNodeFinder->findInstanceOf($stmts, ClassLike::class);
 
-        $reflectionClassName = $classReflection->getName();
         foreach ($classLikes as $classLike) {
-            if ($reflectionClassName !== $desiredClassName) {
+            if (! $this->nodeNameResolver->isName($classLike, $className)) {
```

Comment:
@TomasVotruba this was the bug part, `$reflectionClassName` was always equal with  `$desiredClassName`, so flipped child parent definition was not working.

This patch fix it with compare with classreflection classname, demonstrate at the fixture https://github.com/rectorphp/rector-src/commit/c35e05a08573b39ab34980eafcd8fb86d56fb8e8 👍 🎉